kvbc/test/sparse_merkle_storage/db_adapter_unit_test.cpp

```diff
@@ -183,6 +184,126 @@ TEST_P(db_adapter_custom_blockchain, add_block_return) {
+  const auto numBlocks = 4u;
+  for (auto i = 0u; i < numBlocks; ++i) {
+    ASSERT_EQ(adapter.addBlock(updates), i + 1);
+  }
+
+  for (const auto &k : non_provable_key_set) {
+    const auto [value, blockId] = adapter.getValue(k, numBlocks);
+    ASSERT_EQ(value, defaultSliver);
+    ASSERT_EQ(blockId, numBlocks);
+  }
+
+  const auto [value, blockId] = adapter.getValue(defaultSliver, numBlocks);
+  ASSERT_EQ(value, defaultSliver);
+  ASSERT_EQ(blockId, numBlocks);
+}
+
+TEST_P(db_adapter_custom_blockchain, add_block_only_with_non_provable_keys) {
+  std::string key1{"key1"};
+  std::string key2{"key2"};
+  const Sliver skey1{std::move(key1)};
+  const Sliver skey2{std::move(key2)};
+  concord::kvbc::v2MerkleTree::DBAdapter::NonProvableKeySet non_provable_key_set = {skey1, skey2};
+  auto adapter = DBAdapter{GetParam()->newEmptyDb(), true, non_provable_key_set};
+  ASSERT_THROW(adapter.getValue(skey1, 4), NotFoundException);
+
+  const auto updates = SetOfKeyValuePairs{{skey1, defaultSliver}, {skey2, defaultSliver}};
+  const auto numBlocks = 4u;
+  for (auto i = 0u; i < numBlocks; ++i) {
+    ASSERT_EQ(adapter.addBlock(updates), i + 1);
+  }
+
+  for (const auto &k : non_provable_key_set) {
+    const auto [value, blockId] = adapter.getValue(k, numBlocks);
+    ASSERT_EQ(value, defaultSliver);
+    ASSERT_EQ(blockId, numBlocks);
+  }
+
+  ASSERT_THROW(adapter.getValue(defaultSliver, numBlocks), NotFoundException);
+}
+
+TEST_P(db_adapter_custom_blockchain, raw_block_with_non_provable_keys) {
+  std::string key1{"key1"};
+  std::string key2{"key2"};
+  std::string key3{"key3"};
+  const Sliver skey1{std::move(key1)};
+  const Sliver skey2{std::move(key2)};
+  const Sliver skey3{std::move(key3)};
+  concord::kvbc::v2MerkleTree::DBAdapter::NonProvableKeySet non_provable_key_set = {skey1, skey2, skey3};
+  auto adapter = DBAdapter{GetParam()->newEmptyDb(), true, non_provable_key_set};
+
+  const auto updates = SetOfKeyValuePairs{{skey1, defaultSliver}, {skey2, defaultSliver}, {skey3, defaultSliver}};
+  const auto numBlocks = 4u;
+  for (auto i = 0u; i < numBlocks; ++i) {
+    ASSERT_EQ(adapter.addBlock(updates), i + 1);
+  }
+
+  for (auto i = 0u; i < numBlocks; ++i) {
+    ASSERT_EQ(block::detail::getData(adapter.getRawBlock(i + 1)), updates);
+  }
+}
+
+TEST_P(db_adapter_custom_blockchain, add_block_throws_if_deletes_contain_non_provable_keys) {
+  std::string key1{"key1"};
+  std::string key2{"key2"};
+  const Sliver skey1{std::move(key1)};
+  const Sliver skey2{std::move(key2)};
+  concord::kvbc::v2MerkleTree::DBAdapter::NonProvableKeySet non_provable_key_set = {skey1, skey2};
+  auto adapter = DBAdapter{GetParam()->newEmptyDb(), true, non_provable_key_set};
+
+  const auto updates = SetOfKeyValuePairs{{skey1, defaultSliver}, {skey2, defaultSliver}};
+  const auto deletes = OrderedKeysSet{skey1};
+  ASSERT_THROW(adapter.addBlock(updates, deletes), std::runtime_error);
+}
+
+#ifdef USE_ROCKSDB
```

Comment:
In order to test with both ` memorydb` and `RocksDB`, I think you can use `db_adapter_custom_blockchain` and get the test DB by calling `GetParam()->newEmptyDb()` and keeping the pointer. Then it can be used twice for both DBAdapter instances. That way we can avoid the `ifdef`. Please correct me if I miss something.

Reply:
Great idea, done!